File: ipclass.cpp
#include <iostream>
#include <sstream>
#include <bitset>
#include <vector>

using namespace std;

vector<string> split(const string& s, char delimiter) {
    vector<string> tokens;
    string token;
    istringstream tokenStream(s);
    while (getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}

char findIPAddressClass(const string& ip) {
    vector<string> octets = split(ip, '.');
    int firstOctet = stoi(octets[0]);

    if (firstOctet >= 0 && firstOctet <= 127) {
        return 'A';
    } else if (firstOctet >= 128 && firstOctet <= 191) {
        return 'B';
    } else if (firstOctet >= 192 && firstOctet <= 223) {
        return 'C';
    } else if (firstOctet >= 224 && firstOctet <= 239) {
        return 'D';
    } else if (firstOctet >= 240 && firstOctet <= 255) {
        return 'E';
    } else {
        return 'N'; // Invalid IP address
    }
}

string convertToBinary(const string& ip) {
    vector<string> octets = split(ip, '.');
    string binaryIP;
    for (const string& octet : octets) {
        int value = stoi(octet);
        bitset<8> binaryValue(value);
        binaryIP += binaryValue.to_string();
    }
    return binaryIP;
}

string findNetworkID(const string& ip, char ipAddressClass) {
    vector<string> octets = split(ip, '.');
    if (ipAddressClass == 'A') {
        return octets[0] + ".0.0.0";
    } else if (ipAddressClass == 'B') {
        return octets[0] + "." + octets[1] + ".0.0";
    } else if (ipAddressClass == 'C') {
        return octets[0] + "." + octets[1] + "." + octets[2] + ".0";
    } else {
        return "N/A"; // Network ID not applicable for classes D and E
    }
}

string findFirstIPAddress(const string& networkID, char ipAddressClass) {
    vector<string> octets = split(networkID, '.');
    if (ipAddressClass == 'A' || ipAddressClass == 'B' || ipAddressClass == 'C') {
        octets.back() = "1"; // Set the last octet to 1
        return octets[0] + "." + octets[1] + "." + octets[2] + "." + octets[3];
    } else {
        return "N/A"; // First IP address not applicable for classes D and E
    }
}

string findLastIPAddress(const string& networkID, char ipAddressClass) {
    vector<string> octets = split(networkID, '.');
    if (ipAddressClass == 'A') {
        octets[1] = "255"; // Set the third octet to 255
        octets[2] = "255"; // Set the third octet to 255
        octets[3] = "255"; // Set the third octet to 255
        return octets[0] + "." + octets[1] + "." + octets[2] + "." + octets[3];
    } else if (ipAddressClass == 'B') {
        octets[2] = "255"; // Set the third octet to 255
        octets[3] = "255"; // Set the third octet to 255
        return octets[0] + "." + octets[1] + "." + octets[2] + "." + octets[3];
    } else if (ipAddressClass == 'C') {
        octets[3] = "255"; // Set the last octet to 255
        return octets[0] + "." + octets[1] + "." + octets[2] + "." + octets[3];
    } else {
        return "N/A"; // Last IP address not applicable for classes D and E
    }
}

bool isSuitableIPClass(char ipAddressClass, int numHosts) {
    if (ipAddressClass == 'A') {
        return numHosts <= 16777214; // 2^24 - 2
    } else if (ipAddressClass == 'B') {
        return numHosts <= 65534; // 2^16 - 2
    } else if (ipAddressClass == 'C') {
        return numHosts <= 254; // 2^8 - 2
    } else {
        return false; // Not applicable for classes D and E
    }
}

int main() {
    string ipAddress;
    cout << "Enter an IP address in dotted decimal format: ";
    cin >> ipAddress;

    char ipAddressClass = findIPAddressClass(ipAddress);
    if (ipAddressClass != 'N') {
        cout << "IP address class: " << ipAddressClass << endl;
        
        int numHosts;
        cout << "Enter the number of hosts required: ";
        cin >> numHosts;

        if (isSuitableIPClass(ipAddressClass, numHosts)) {
            string networkID = findNetworkID(ipAddress, ipAddressClass);
            if (networkID != "N/A") {
                cout << "Network ID: " << networkID << endl;
                string firstIPAddress = findFirstIPAddress(networkID, ipAddressClass);
                string lastIPAddress = findLastIPAddress(networkID, ipAddressClass);
                if (firstIPAddress != "N/A") {
                    cout << "First IP address: " << firstIPAddress << endl;
                    cout << "Last IP address: " << lastIPAddress << endl;
                } else {
                    cout << "First IP address: N/A" << endl;
                    cout << "Last IP address: N/A" << endl;
                }
            } else {
                cout << "Network ID: N/A" << endl;
            }
        } else {
            cout << "The IP class is not suitable for the given number of hosts." << endl;
        }
    } else {
        cout << "Invalid IP address!" << endl;
    }

    return 0;
}

File: ipclass.py
import ipaddress

def convert_to_binary(ip_address):
    octets = ip_address.split('.')
    binary_ip = []
    for octet in octets:
        binary_octet = bin(int(octet))[2:].zfill(8)
        binary_ip.append(binary_octet)
    return '.'.join(binary_ip)

def get_ip_class(ip_address):
    first_octet = int(ip_address.split('.')[0])
    if 1 <= first_octet <= 126:
        return 'Class A'
    elif 128 <= first_octet <= 191:
        return 'Class B'
    elif 192 <= first_octet <= 223:
        return 'Class C'
    elif 224 <= first_octet <= 239:
        return 'Class D (Multicast)'
    elif 240 <= first_octet <= 255:
        return 'Class E (Reserved)'

def is_valid_hosts(ip_address, num_hosts):
    network = ipaddress.IPv4Network(ip_address)
    max_hosts = network.num_addresses - 2  # Subtract 2 for network and broadcast addresses
    return 0 <= num_hosts <= max_hosts

def get_network_range(ip_address, num_hosts):
    network = ipaddress.IPv4Network(ip_address)
    network_address = str(network.network_address)
    last_ip = str(network.network_address + num_hosts + 1)
    return network_address, last_ip

# Example usage
ip = input("Enter an IP address in dotted decimal format: ")
num_hosts = int(input("Enter the number of hosts: "))

binary_ip = convert_to_binary(ip)
print("Binary IP address:", binary_ip)

ip_class = get_ip_class(ip)
print("IP address class:", ip_class)

if is_valid_hosts(ip, num_hosts):
    first_ip, last_ip = get_network_range(ip, num_hosts)
    print("First IP of the network:", first_ip)
    print("Last IP of the network:", last_ip)
else:
    print("Invalid number of hosts for the given IP address.")
